p2p/discovery/message.go

```diff
@@ -18,39 +20,189 @@ const (
 	NeighborsMsg MsgType = 5
 )
 
-type Ping struct {
+const (
+	VERSION uint = 1
+)
+
+type ping struct {
 	Version uint // TODO add version check
-	ID      NodeID
+	SelfID  NodeID
+
+	to NodeID
 }
 
-type Pong struct {
-	ID NodeID
+type pong struct {
+	SelfID NodeID
 }
 
-type FindNode struct {
-	Target NodeID
+type findNode struct {
+	SelfID  NodeID
+	QueryID NodeID
+
+	to NodeID
 }
 
-type Neighbors struct {
-	Nodes []RPCNode
+type neighbors struct {
+	SelfID NodeID
+	Nodes  []*rpcNode
 }
 
-type RPCNode struct {
-	ID      NodeID
+type rpcNode struct {
+	SelfID  NodeID
 	IP      net.IP
 	UDPPort uint16
 }
 
-func ByteToMsgType(byte byte) MsgType {
+func (r *rpcNode) ToNode() *Node {
+	return NewNode(r.SelfID, r.IP, r.UDPPort)
+}
+
+func byteToMsgType(byte byte) MsgType {
 	return MsgType(uint8(byte))
 }
 
-func MsgTypeToByte(t MsgType) byte {
+func msgTypeToByte(t MsgType) byte {
 	return byte(t)
 }
 
 func generateBuff(code MsgType, encoding []byte) []byte {
-	buff := []byte{MsgTypeToByte(code)}
+	buff := []byte{msgTypeToByte(code)}
 
 	return append(buff, encoding...)
 }
+
+// handle send ping msg and add pending
+func (m *ping) handle(t *udp, from *net.UDPAddr) {
+	log.Debug("received ping from: %s", common.BytesToHex(m.SelfID.Bytes()))
+
+	// response with pong
+	if m.Version != VERSION {
```

Comment:
you hardcode the version here, how can we do version compatible and multiple version co-existence?

Reply:
it's not hard code, it checks